File: lua/SpringBox.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt
if (not SpringBox) then

SpringBox				=	{}
SpringBox.className	    =	"SpringBox"
SpringBox.ZEROVEL_EPS	=	0.01

    function SpringBox.Create(startx,y,endx,m,r,k,time)

        local springbox = {
            osc             =   KDampedOscillator.Create( m or 1,  r or 12,  k or 100,0,0),
            osctimer        =	Ktimer.Create(1,true),
            y               =   y,
			time			= 	0,
			finishby		=	time or 5,
            className       =   SpringBox.className,
        }
        setmetatable(springbox,{__index = SpringBox } )
		springbox.osc:reset( endx,  0,  startx)
        return springbox
    end
	
	function SpringBox.Start(springbox)
		Ktimer.resume(springbox.osctimer)
	end
	

    function SpringBox.IsFinished(springbox)
        return (springbox.finishby and springbox.time > springbox.finishby)  
    end

	function SpringBox.IsStationary(springbox)
		 return KDampedOscillator.equilibrium(springbox.osc,Ktimer.elapsed(springbox.osctimer),SpringBox.ZEROVEL_EPS)
	end

    function SpringBox.Update(springbox,dt)
		springbox.time	=	springbox.time + dt
    end

    function SpringBox.GetPosition(springbox)
        local elapsed = Ktimer.elapsed(springbox.osctimer)
        local cx = KDampedOscillator.evaluate(springbox.osc, elapsed)
        return { x = cx, y =  springbox.y }
    end

end

File: source/library/SpringBox.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

function SpringBox(startx,y,x,m,r,k,time)
{
	var darg = function(val,defval) { return (typeof(val) === 'undefined' || typeof(val) !== typeof(defval) ? defval : val) ; }

 	this.osc          	=   KDampedOscillator.Create( darg(m,1),  darg(r,12),  darg(k,100),0,0) ;
	this.osctimer      	=	Ktimer.Create(1,true) ;
	this.endx			=	x ;
    this.y				=   y ;
	this.startx			=	startx ;
	this.time			= 	0 ;
	this.finishby		=	darg(time,5) ;
	this.started		=	false ;
	this.osc.reset( this.endx,  0,  this.startx) ;
	return this ;
}

SpringBox.className	    =	"SpringBox"
SpringBox.ZEROVEL_EPS	=	0.1

SpringBox.Create = function(startx,y,x,m,r,k,time)
{
	return new SpringBox(startx,y,x,m,r,k,time) ;
}

SpringBox.prototype = {

	constructor: SpringBox,

	Start: function()
	{
		this.started	=	true ;
		this.osctimer.resume() ;
	},
	
	GetDuration: function()
	{
		return this.finishby ;
	},
	
	
    IsFinished: function()
	{
        return (this.time > this.finishby)  ;
    },

	IsStationary: function()
	{
		 return this.osc.equilibrium(this.osctimer.elapsed(),SpringBox.ZEROVEL_EPS) ;
	},

    Update: function(dt)
	{
		if (!this.IsFinished() && this.started) 
		{
			this.time	=	this.time + dt ;
		}
    },
	

    GetPosition: function()
	{
    	var elapsed = this.osctimer.elapsed()
   		var cx = this.osc.evaluate(elapsed)
        return { x: (this.IsStationary() ? this.startx : parseInt(cx)), y: this.y } ;

    },
	
	toString: function()
	{
		return "SpringBox" ;
	}
	

}
